Test Command (#236)
sourcegraph/scip: docs/CLI.md

   --comment-syntax value  Comment syntax to use for snapshot files (default: "//")
```

## `scip test`

```
NAME:
   scip test - Validate subsets of snapshot files

USAGE:
   scip test [command options] [arguments...]

DESCRIPTION:
   The test subcommand validates whether a provided scip index
   file matches manually specified symbol fields within an index. This can be used in
   conjunction with snapshot to provide a cohesive testing framework for scip indexers.

OPTIONS:
   --from value            Path to SCIP index file (default: "index.scip")
   --comment-syntax value  Comment syntax to use for snapshot files (default: "//")
   --help, -h              show help
```

> ### File Specification

Author: Looks like this file `docs/CLI.md` is used in a unit test to validate the output of `scip --help`

I'm open to suggestions on a better location for this documentation

Reviewer: I realize I recommended linking to the `main` branch earlier, but it's probably better to link to the URL for the released version so that there is less risk of confusion in case an old version of the subcommand doesn't support some syntax supported by a new version/on main. I've pushed a change for that.

   --comment-syntax value  Comment syntax to use for snapshot files (default: "//")
```

## `scip test`

```
NAME:
   scip test - Validate subsets of snapshot files

USAGE:
   scip test [command options] [arguments...]

DESCRIPTION:
   The test subcommand validates whether a provided scip index
   file matches manually specified symbol fields within an index. This can be used in
   conjunction with snapshot to provide a cohesive testing framework for scip indexers.

OPTIONS:
   --from value            Path to SCIP index file (default: "index.scip")
   --comment-syntax value  Comment syntax to use for snapshot files (default: "//")
   --help, -h              show help
```

### File Specification

> The validation files for `scip test` are loosely based on [sublime text's](https://www.sublimetext.com/docs/syntax.html#testing) syntax highlighting files. Symbol ranges are selected using `^`, `^^` and `// <-`. Symbol attributes can be specified as the first word: `definition`, `reference`, `diagnostic`, and all subsequent lines are interpreted based on the kind of symbol attribute you are writing a test for

Reviewer: ```suggestion
The validation files for `scip test` are loosely based on [Sublime Text's](https://www.sublimetext.com/docs/syntax.html#testing) syntax highlighting files. Symbol ranges are selected using `^`, `^^` and `// <-`. Symbol attributes can be specified as the first word: `definition`, `reference`, `diagnostic`, and all subsequent lines are interpreted based on the kind of symbol attribute you are writing a test for
```

Here, `diagnostic` is technically a field and definition/reference are roles. 'symbol attributes' sounds like standard SCIP terminology, but it actually doesn't have any meaning in SCIP. Later in this doc, the same thing is referred to as 'selection kind'. Alternate terminology suggestion, say something like:

> There are 2 types of checks supported by `scip test` - for symbol roles and for diagnostics. ...
>
> The syntax to check symbol roles is as follows: \<code snippet\>
>
> The syntax to check diagnostics is as follows: \<code snippet\>

If you want, it might make sense to add a way to check/test relationships too.

Given the following example, the cli command `scip test` will exit with a non-zero exit code if the provided `index.scip` file does not contain a definition, at the specified range, with the specified symbol
```js
function someFunction() {
   //    ^^^^^^^^^^^^ definition scip-typescript npm test_package 1.0.0 lib/`test.js`/someFunction().
}
```

Three range selection comment formats are supported:

> - `// ^^^`: enforces the length of the symbol. Will fail if the range at this location does not equal 3 characters

Reviewer: ```suggestion
- `// ^^^`: enforces the length of the occurrence. Will fail if the range at this location does not equal 3 characters
```

> - `// ^`: ignore length, symbol can occur at any point between the start and end of the `^` character

Reviewer: ```suggestion
- `// ^`: ignore length, `^` can occur at any point to any character in the occurrence
```

> - `// <-`: ignore length, and treat the "start" of the symbol at the beginning of the comment. This follows sublime-text's format

Reviewer: ```suggestion
- `// <-`: ignore length, and treat the character above the first comment character as the start of the occurrence, similar to Sublime Text
```
